include/odp/api/spec/packet_io.h

```diff
@@ -990,6 +1069,171 @@ uint64_t odp_pktin_wait_time(uint64_t nsec);
+	 * segment transmission.
+	 */
+	uint32_t payload_offset;
+
+	/** LSO profile handle */
+	odp_lso_profile_t lso_profile;
+
+} odp_pktout_lso_t;
+
+/**
+ * Initialize LSO profile parameters
+ *
+ * Initialize an odp_lso_profile_param_t to its default values for all fields.
+ *
+ * @param param Address of the odp_lso_profile_param_t to be initialized
+ */
+void odp_lso_profile_param_init(odp_lso_profile_param_t *param);
+
+/**
+ * Create LSO profile
+ *
+ * LSO profile defines the set of segmentation operations to be performed to a packet. LSO profiles
+ * are created before the packet IO interface is started (after odp_pktio_config() and before
+ * odp_pktio_start()).
+ *
+ * @param pktio   Packet IO interface which is used with this LSO profile
+ * @param param   LSO profile parameters
+ *
+ * @return LSO profile handle
+ * @retval ODP_LSO_PROFILE_INVALID on failure
+ */
+odp_lso_profile_t odp_lso_profile_create(odp_pktio_t pktio, const odp_lso_profile_param_t *param);
+
+/**
+ * Destroy LSO profile
+ *
+ * LSO profiles can be destoyed only when the packet IO interface is not active (i.e. after it
+ * has been stopped).
+ *
+ * @param lso_profile   LSO profile to be destroyed
+ *
+ * @retval 0 on success
+ * @retval <0 on failure
+ */
+int odp_lso_profile_destroy(odp_lso_profile_t lso_profile);
+
+/**
+ * Send packets with segmentation offload
+ *
+ * Like odp_pktout_send(), but splits a packet payload into 'seg_payload' or smaller segments
+ * during output. Packet headers (before 'payload_offset') are copied into each segment and
+ * automatically modified before transmission. Header updates are based on the defined L3 and L4
+ * protocols. Header checksums are updated after modifications. L3/L4 protocol modifications
+ * require that L3/L4 layer offsets in the packet are valid (see e.g. odp_packet_l3_offset()).
+ * Segmentation is done on the requested layer (L3, L4 or none).
+ *
+ * In addition, custom field updates may be used to cover unsupported or proprietary protocols.
+ * Custom fields must not overlap with each other, or with L3/L4 fields to be updated.
+ *
+ * Packet are processed and transmitted in the array order. Segments of each packet are transmitted
+ * in ascending order.
+ *
+ * Check LSO support level from packet IO capabilities (odp_pktio_capability_t).
+ *
+ * @param queue        Packet output queue handle
+ * @param packet[]     Array of packets to send
+ * @param lso_opt[]    Array of packet segmentation options
+ * @param num_pkt      Number of packets to send
+ * @param num_opt      Number of options. Valid values are:
+ *                       0: no options, LSO is not performed
+ *                       1: single option applies for all packets
+ *                       num_pkt: an option per packet
+ *
+ * @return Number of packets successfully segmented (0 ... num_pkt)
+ * @retval <0 on failure
+ */
+int odp_pktout_send_lso(odp_pktout_queue_t queue, const odp_packet_t packet[],
```

Comment:
Why shouldn't we have LSO offload similiar to l3/l4 checksum offload in Tx to be configured via odp_pktio_config()(with profiles) so that we can use all odp_pktout_send(), odp_queue_enq(), odp_tm_enq() with this offload.

For odp_pktout_lso_t opts, we can have odp_packet_*() API's similar to checksum insert API's to pass these options per packet before pktout.

By defining new API for LSO in Direct mode, we have to have new API's in other modes and with TM.

Not sure if it was already discussed but just wanted to share my thoughts here.

Reply:
Checksum offload requires only one bit of additional metadata, whereas LSO needs metadata for frag len, payload offset and profile handle. I have tried to avoid metadata expansion, but we may need to go that route when LSO needs to be combined with async send (queue enqueue) and TM.

Payload offset could be generally usable metadata field (set by parser after e.g. successfully L4 parse). With that in mind additional metadata for LSO would be only two fields.

Reply:
Agreed. Only two more meta fields one for lso profile that also implies lso enabled and second for max_seg_payload